### ISSUE TITLE
Environment variables from secrets are not exported on OCP 3.11

### ISSUE THREAD

**Issue by mjurc** [quarkusio/quarkus#9561]
**Describe the bug**
Exporting secrets as environment variables as described by [Deploying to OpenShift guide](https://quarkus.io/guides/deploying-to-openshift#environment-variables-from-secret) does not work on OCP 3.11.

The same feature works correctly on OCP 4.x.

**Expected behavior**
Secrets should be present in environment variables specified by the property name.

**Actual behavior**
Environment variables are not available in deployment pods on OCP 3.11.

**To Reproduce**
1. Checkout https://github.com/mjurc/mock-artifacts  
2. `oc login` to OCP 3.11
3. `cd quarkus/getting-started/`
4. `oc apply -f src/main/resources/secret.yaml`
5. `mvn clean package`
6. check the deployment pod's env vars - the environment variables are not exported in environment variables as defined by the application's `application.properties` file, e. g. `DB_DATABASE=postgresql`

**Configuration**
```properties
quarkus.openshift.expose=true
quarkus.kubernetes-client.trust-certs=true
quarkus.s2i.base-jvm-image=registry.access.redhat.com/openjdk/openjdk-11-rhel7

quarkus.openshift.env-vars.DB_DATABASE.secret=postgresql
quarkus.openshift.env-vars.DB_DATABASE.value=database
quarkus.openshift.env-vars.DB_USERNAME.secret=postgresql
quarkus.openshift.env-vars.DB_USERNAME.value=username
quarkus.openshift.env-vars.DB_PASSWORD.secret=postgresql
quarkus.openshift.env-vars.DB_PASSWORD.value=password
```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `Linux tigris 5.6.13-300.fc32.x86_64 #1 SMP Thu May 14 22:51:37 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
 - Output of `java -version`: 
```
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.6+10, mixed mode)
```
 - GraalVM version (if different from Java): `N/A` 
 - Quarkus version or git rev: 1.3.y, 1.4.2
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): `Apache Maven 3.6.2 (40f52333136460af0dc0d7232c0dc0bcf0d9e117; 2019-08-27T17:06:16+02:00)`

**Additional context**
(Add any other context about the problem here.)

**Comment by geoand** [comment]
@iocanel @maxandersen have you seen this kind of discrepancy between OCP 3.11 and 4.x?

**Comment by maxandersen** [comment]
I haven't tried that specific setup in 3.11 - surprised there would be a difference here though...

**Comment by geoand** [comment]
Yeah, me too....

**Comment by iocanel** [comment]
That's weird. This is something that we do test.
I am wondering if the problem lies in `DB_PASSWORD` and how the quarkus config framework handles capitals and/or underscores.

**Comment by geoand** [comment]
Do you plan to check it out @iocanel ?

**Comment by iocanel** [comment]
I think that the problem here lies in the expectation. 

While the quarkus kubernetes extension does support passing environment variables from secret, it doesn't provide fine-grained control. This means that the whole secret is passed instead of adding property by property.

So the following configuration:

```
quarkus.openshift.env-vars.DB_DATABASE.secret=postgresql
quarkus.openshift.env-vars.DB_DATABASE.value=database
quarkus.openshift.env-vars.DB_USERNAME.secret=postgresql
quarkus.openshift.env-vars.DB_USERNAME.value=username
quarkus.openshift.env-vars.DB_PASSWORD.secret=postgresql
quarkus.openshift.env-vars.DB_PASSWORD.value=password
```

could be replaced with:

```
quarkus.openshift.env-vars.THIS_IS_IGNORED.secret=postgresql
```

Or using the new style:

```
quarkus.kubernetes.env.secrets=postgres
```

**Comment by mjurc** [comment]
@iocanel Thanks for looking into this. 

I still don't see `DB_DATABASE`, `DB_USERNAME` or `DB_PASSWORD` on OCP3.11 holding the secret, though. The values respectively are `database`, `username` and `password. If I remove the `*.value=*` lines from yaml, the environment variables are not exported at all. Shouldn't the contain the secret?

Is it possible to configure the export of single key/value pair into an environment variable in `application.properties`? How would the example above look, if it was?

Is the secrets being exported as the vars just a fluke on OCP4 side?

**Comment by iocanel** [comment]
> @iocanel Thanks for looking into this.
> 
> I still don't see `DB_DATABASE`, `DB_USERNAME` or `DB_PASSWORD` on OCP3.11 holding the secret, though. The values respectively are `database`, `username` and `password`. If I remove the `*.value=*` lines from yaml, the environment variables are not exported at all. Shouldn't the contain the secret?

By having just:

```
quarkus.openshift.env-vars.THIS_IS_IGNORED.secret=postgresql

```

or 

```
quarkus.kubernetes.env.secrets=postgresql
```

The generated YAML will contain:

```
       envFrom:
        - secretRef:
            name: postgresql
```

And yes, this doesn't allow you to do any sort of mapping, like: `database -> DB_DATABASE`. The environment variable will be named `database` or whatever is the entry in your secret.

> 
> Is it possible to configure the export of single key/value pair into an environment variable in `application.properties`? How would the example above look, if it was?

If you mean to just add a single key/value pair from the secret, then no, its not possible ATM!

> Is the secrets being exported as the vars just a fluke on OCP4 side?

The generator is unaware of your cluster version. The generated yaml, is expected to be the same regardless of the target cluster, it something that can work even offline.

**Comment by mjurc** [comment]
@iocanel Alright, I believe I found out what's the problem. As you said, Quarkus does not know about where it is going to be deployed. The `env` section from generated `kubernetes.yaml` is as follows:

```
      - env:
        - name: "KUBERNETES_NAMESPACE"
          valueFrom:
            fieldRef:
              fieldPath: "metadata.namespace"
        - name: "DB_DATABASE"
          valueFrom:
            secretKeyRef:
              key: "database"
              name: "postgresql"
              optional: false
        - name: "DB_PASSWORD"
          valueFrom:
            secretKeyRef:
              key: "password"
              name: "postgresql"
              optional: false
        - name: "DB_USERNAME"
          valueFrom:
            secretKeyRef:
              key: "username"
              name: "postgresql"
              optional: false
```
`secretKeyRef` does not have `optional` field on OCP 3.11 as far as I can see - the secret works if the `optional` fields are removed.

**Comment by geoand** [comment]
@iocanel I am little confused on what the action item is on our side for this. Can you elaborate please?

**Comment by iocanel** [comment]
It seems that I am wrong in that mapping secret and configmap keys to environment variables was not expected to work. It seems that it did and now it is broken.

This issue is a duplicate of https://github.com/quarkusio/quarkus/issues/9629 and  I suggest continuing the discussion there.

**Comment by geoand** [comment]
Hm... But this issue also affects `1.3.x`, right @mjurc?

**Comment by gsmet** [comment]
Is this one fixed or not?

**Comment by geoand** [comment]
It was fixed on the 1.3 branch. @iocanel were you expecting @metacosm to fix it for `master`?

**Comment by iocanel** [comment]
The fix could be applied to master as is, but tests might not be transferable due to #9629. But I am not 100% sure of that. 

I will give it a shot and let you know.
